app/controllers/BalanceRequestController.scala

```diff
@@ -72,61 +88,151 @@ class BalanceRequestController @Inject() (
       }
 
-  def submitBalanceRequest: Action[BalanceRequest] =
-    authenticate().io(parse.json[BalanceRequest]) { implicit request =>
+  private def validateTaxIdentifier(
+    taxId: TaxIdentifier
+  ): ValidatedNel[BadRequestError, TaxIdentifier] = {
+    (
+      Validated.condNel(
+        taxId.value.length <= 17,
+        taxId,
+        InvalidTaxIdentifier(reason = "Tax identifier value has a maximum length of 17 characters")
+      ),
+      Validated.condNel(
+        taxId.value.forall(_.isLetterOrDigit),
+        taxId,
+        InvalidTaxIdentifier(reason = "Tax identifier value must be alphanumeric")
+      )
+    ).tupled.as(taxId)
+  }
+
+  private def validateGuaranteeReference(
+    grn: GuaranteeReference
+  ): ValidatedNel[BadRequestError, GuaranteeReference] = {
+    (
+      Validated.condNel(
+        grn.value.length >= 17,
+        grn,
+        InvalidGuaranteeReference(reason =
+          "Guarantee reference value has a minimum length of 17 characters"
+        )
+      ),
+      Validated.condNel(
+        grn.value.length <= 24,
+        grn,
+        InvalidGuaranteeReference(reason =
+          "Guarantee reference value has a maximum length of 24 characters"
+        )
+      ),
+      Validated.condNel(
+        grn.value.forall(_.isLetterOrDigit),
+        grn,
+        InvalidGuaranteeReference(reason = "Guarantee reference value must be alphanumeric")
+      )
+    ).tupled.as(grn)
+  }
+
+  private def validateAccessCode(
+    code: AccessCode
+  ): ValidatedNel[BadRequestError, AccessCode] = {
+    (
+      Validated.condNel(
+        code.value.length == 4,
+        code,
+        InvalidAccessCode(reason = "Access code value must be 4 characters in length")
+      ),
+      Validated.condNel(
+        code.value.forall(_.isLetterOrDigit),
+        code,
+        InvalidAccessCode(reason = "Access code value must be alphanumeric")
+      )
+    ).tupled.as(code)
+  }
+
+  private def validateBalanceRequest(
+    request: BalanceRequest
+  ): ValidatedNel[BadRequestError, BalanceRequest] = {
+    (
+      validateTaxIdentifier(request.taxIdentifier),
+      validateGuaranteeReference(request.guaranteeReference),
+      validateAccessCode(request.accessCode)
+    ).tupled.as(request)
+  }
+
+  private def validateRequest(
+    continue: BalanceRequest => IO[Result]
+  )(implicit request: Request[JsValue]): IO[Result] =
+    IO(request.body.validate[BalanceRequest]).flatMap {
+      case JsError(errors) =>
+        val error     = JsonParsingError(errors = errors)
+        val errorJson = Json.toJson(error)
+        IO.pure(BadRequest(errorJson))
+      case JsSuccess(balanceRequest, _) =>
+        validateBalanceRequest(balanceRequest)
+          .map(continue)
+          .valueOr { errors =>
+            val error     = MultipleErrors(errors = errors)
+            val errorJson = Json.toJson[BadRequestError](error)
+            IO.pure(BadRequest(errorJson))
+          }
+    }
```

Comment:
See the above. Plus its a lot of noise to add to the controller that is hard to test the behaviour of

Reply:
I've factored this validation out into a service now - let me know what you think!